File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-2.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        window = set()
        l,r = 0, 0
        res=0
        while l<=r and r<len(s):
            while window and s[r] in window:
                window.remove(s[l])
                l+=1
            window.add(s[r])
            res=max(res, r-l+1)
            r+=1
        return res

File: Data Structures & Algorithms/eating-bananas/submission-2.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        //  h= time to eat all bananas
        // min speed to eat all bananas within h hours
        int l=0;
        int r=Arrays.stream(piles).max().getAsInt();
        int res=r;
        while(l<=r){
            int mid = (r-l)/2+l;
            long timeTaken=0;
            for(int pile : piles){
                timeTaken+=Math.ceil((double)pile/mid);
            }
            if(timeTaken<=h){
                res=mid;
                r=mid-1;
            }else{
                l=mid+1;
            }
        }

        return res;
    }
}
